docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+```wing
+bring clients;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  // version: "0.0.0",
+  // outdir: join(@dirname, "dist"),
+);
+```
+
+To lift a resource into the client to make it accessible to consumers, add a `lift` expression:
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+client.lift(lift { bucket: [put] });
+```
+
+In order to access the bucket, the client needs to know it should resolve the relevent information (e.g. bucket name or handle).
+This information is stored in client.context, so we need to way to both store and resolve this information.
+For example, we can store/retrieve the context from a file:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return fs.readJson("/var/run/wing-client-context.json");
+  },
+);
+
+// The consumer of the client will need to retrieve/link this data after deployment
+let clientContext = new cloud.Bucket() as "ctx";
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+If the runtime cost is acceptable, the context can even be directly resolved directly from the bucket during the client's runtime:
+
+```wing
+bring cloud;
+bring clients;
+
+let clientContext = new cloud.Bucket() as "ctx";
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    return Json.parse(clientContext.get("ctx"));
+  },
+);
+
+// hydrates the context during deployment
+clientContext.addObject("ctx", Json.stringify(client.context))
+```
+
+In cases where reading from the filesystem is difficult and the overhead or a runtime lookup is too high, the context can be passed in via an environment variable:
+
+```wing
+bring cloud;
+bring clients;
+bring util;
+
+let bucket = new cloud.Bucket() as "shared";
+
+let client = new clients.TypeScript(
+  name: "@acme/infra-client",
+  contextResolver: inflight () => {
+    let ctxRaw = util.getEnv("WING_CLIENT_CONTEXT");
+    let ctxDecoded = util.fromBase64(ctxRaw);
+    return Json.parse(ctxDecoded);
+  },
+);
+
+client.add(lift { bucket: [put] });
+```
+
+Regardless of the approach used to resolve the context, using the generated client will work the same way:
+
+```ts
+import { Client } from "@acme/infra-client";
+
+// performs the context resolution
+const client = new Client();
+
+await client.bucket.put("key", "value");
+```
+
+The client code is the same as the generated javascript as if it were used within wing itself!
+
+## Implementation Details
+
+### Client context data
+
+The client context data is a JSON object that contains all the necessary information to connect to the resources.
+For now, the only information that needs to be stored are environment variables.
+This corresponds to the only API available on `IInflightHost`, which is `addEnvironment`, which is how these values are added:
```

Comment:
This RFC should discuss changing IInflightHost and and also all inflight clients? 
They currently use environment variables, and this proposal was just going to reuse those inflights.

I understand what you mean, the way clients work now does not need to rely completely on env, but I didn't think we needed to change all that here in order for a TypeScriptClient to work.

Reply:
Mmmmm. I see, so the intent is that when a client is instantiated by the consuming app, it will mutate the current process' environment so that the inflight clients can be initialized.

I think we should change our implementation of `cloud.Function` (and any other inflight host) to use `TypeScriptClient` under the hood, so there's actually no real value in keeping the current environment variables.

But tactically, I can see the incremental value. Let's make sure this is discussed in the RFC.

Reply:
> Let's make sure this is discussed in the RFC.

I don't see how this is in scope of the RFC. This RFC is a proposal within an architecture we already have. I don't disagree with your point about environment variables but I feel like the discussion in these threads is a sufficient record and any desire for a wider change can be in its own PR.

Reply:
I am okay to punt this, but just want to make sure we are aware that we are introducing an abstraction leak and that's debt we should clean up.

Reply:
This leak is not exposed to the user, the context value is opaque. I'm writing about it explicitly here because this porting of the RFC is discussing implementation details.